feat(schedules schema):added ScheduleRulest, ScheduleFixedInterval

tanushree04 opened the pull request.

tanushree04 committed:
Added Schedule File, Schedule Rule set

tanushree04 committed:
Update ScheduleRuleset.py

tanushree04 committed:
Schedule Day added as List

tanushree04 committed:
Merge branch 'pr-Schedules' of github.com:ladybug-tools-in2/energy-model-schema into pr-Schedules

tanushree04 committed:
Update ScheduleFile.py

tanushree04 committed:
Update ScheduleFile.py

chriswmackey reviewed (commented):
Thanks for putting this together @tanushree04 .  I have a few comments.

chriswmackey commented on app/models/energy/ScheduleFile.py:
```diff
@@ -0,0 +1,57 @@
+
+    values: List[int] = Schema(
+        ...,
+        minItems=8760,
```
I believe that the min here should be 24 since, if someone makes a schedule with start day of 365 on a year that is not a leap year, the number of values in the list should be 24 for the single day of the year.

tanushree04 replied:
That makes sense!

macumber replied:
There are some limitations in E+.  Approach 1 is to accept CSV from user and then generate a compliant CSV to pass to E+.  Approach 2 is to limit your users to valid E+ CSVs.

https://github.com/NREL/EnergyPlus/issues/4154
https://github.com/NREL/EnergyPlus/issues/6898

chriswmackey replied:
@macumber ,
I had some discussions with @mostaphaRoudsari about this and we are in favor of something closer to approach 1, where we have the user send a start day of the year and the minimum number of values needed to define the schedule for the simulation RunPeriod. These two parameters can be used to create the E+ compliant CSV by simply adding filler values for the unimportant hours.  Also, we are sure that we would rather have the user specify the schedule values within the JSON and have the energy_model_measure generate the compliant CSVs rather than have users send separate CSV files that we then have to validate in another script.

chriswmackey commented on app/models/energy/ScheduleRuleset.py:
```diff
@@ -0,0 +1,207 @@
+    numeric_type:  Union[ScheduleContinuous, ScheduleDiscrete]
+
+
+class UnitType (str, Enum):
```
Should this class be repeated for both the `ShceduleFile` and the `ScheduleRuleset`?  I feel that there should be some way to have these two reference the same object.

tanushree04 replied:
I have made changes, so that `UnitType` is being referenced in `ScheduleFile`.

macumber replied:
Common way to do this is factor shared content into a third file, maybe `ScheduleBase` or `ScheduleCommon`

macumber commented on app/models/energy/ScheduleRuleset.py:
```diff
@@ -0,0 +1,207 @@
+
+
+class ScheduleTypeLimits (BaseModel):
+    """Specifies the data types and limits for values contained in schedules."""
```
Are you going to implement custom schema validators which enforce these limits?

tanushree04 replied:
Should I add custom validators that ensure `unit_type` such as `Temperature`, `ActivityLevel` have the correct `numeric_type` -`continuous` is assigned etc.?

macumber replied:
That seems good to me, @chriswmackey do you think that is overkill?

chriswmackey replied:
I agree that I don't think that we need custom validators and it would be overkill

macumber commented on app/models/energy/ScheduleRuleset.py:
```diff
@@ -0,0 +1,207 @@
+        regex=r'^[\s.A-Za-z0-9_-]*$'
+    )
+
+    schedule_rule_set: List[ScheduleRuleset]
```
I think it is better to have `ScheduleRuleset` contain a list of `ScheduleRule` objects.

macumber replied:
Rules are not shared between rulesets

macumber commented on app/models/energy/ScheduleRuleset.py:
```diff
@@ -0,0 +1,207 @@
+        le=31
+    )
+
+    specific_month: int = Schema(
```
What is specific month?

chriswmackey replied:
... and, along the same lines, what is `specific_day`?

tanushree04 replied:
The openstudio IDD file allows the `Date Specification Type` to be either a `DateRange` or `SpecificDates`. A schedule can be assigned to a list of specific dates using `SpecificDates` . The front end of OpenStudio however only allows a `DateRange` to be specified. 
Should I still keep the `SpecificDates` option in the schema? @chriswmackey @mostaphaRoudsari

chriswmackey replied:
I don't think that we need SpecificDates in the schema.

macumber commented on app/models/energy/ScheduleRuleset.py:
```diff
@@ -0,0 +1,207 @@
+
+    date_specification_type: DateSpecificationType = DateSpecificationType.date_range
+
+    start_month: int = Schema(
```
I think it would be better to use a real date type here.  @chriswmackey @mostaphaRoudsari how do you feel about requiring all dates be fully qualified dates with years?

chriswmackey replied:
That seems like a good idea but we have to think about how it should be implemented.  If we have date objects, they will likely get used in a lot of different places and it's better to have just one rather than separate copies for E+ schedules, Annual daylight schedules, etc.  This is why we have date objects within ladybug core that get used by a lot of different parts of the insects.  Let's discuss this more in the meeting today.

tanushree04 committed:
Initial instances and tests for schedules.

tanushree04 committed:
Initial instances and tests schedules

tanushree04 committed:
Added Schedule to Schema

tanushree04 committed:
Added Schedule to Schema

mostaphaRoudsari committed:
fix(model-schema): Construction -> construction

@tanushree04, this should fix the current issue with the failing test.

mostaphaRoudsari committed:
fix(tests): Construction > construction

Another try to fix this one: https://travis-ci.org/ladybug-tools-in2/energy-model-schema/jobs/543196054#L594

macumber commented on app/models/energy/ScheduleBase.py:
```diff
@@ -0,0 +1,60 @@
+from datetime import datetime
+
+
+class UnitType (str, Enum):
```
Seems like this should be in a module or something to make it specific to schedules.  Either Schedule::UnitType or ScheduleUnitType

macumber commented on app/models/energy/ScheduleBase.py:
```diff
@@ -0,0 +1,60 @@
+    control = 'Control'
+    mode = 'Mode'
+
+class DateTime(BaseModel):
```
This was going to move to a more generic location right?

macumber replied:
Maybe break into a separate Date and Time classes with DateTime being a combination?

tanushree04 replied:
@macumber I created a `ScheduleBase` , the `ScheduleFile` and `ScheduleRuleset` reference this file.

macumber commented on app/models/energy/ScheduleFile.py:
```diff
@@ -0,0 +1,57 @@
+
+    type: Enum('ScheduleFile', {'type': 'ScheduleFile'})
+
+    name: str = Schema(
```
This is just a name?  Not a filename?

chriswmackey replied:
I was imagining that the file name will be generated from the schedule name in the energy-model-measure since the current plan is to have the measure write all of the values into the file.

chriswmackey replied:
This is no longer relevant since we decided to go with ScheduleInterval.

chriswmackey replied:
This isn't relevant anymore now that we have decided to go with fixed interval schedules.

macumber commented on app/models/energy/ScheduleFile.py:
```diff
@@ -0,0 +1,57 @@
+
+    values: List[int] = Schema(
+        ...,
+        minItems=24,
```
This looks like it might translate to [ScheduleFixedInterval](https://github.com/NREL/OpenStudio/blob/develop/openstudiocore/src/model/ScheduleFixedInterval.hpp) rather than [ScheduleFile](https://github.com/NREL/OpenStudio/blob/develop/openstudiocore/src/model/ScheduleFile.hpp) in OpenStudio.  ScheduleFixedInterval has all the values embedded in the OSM where ScheduleFile refers to an external file.

chriswmackey replied:
I didn't realize that is what ScheduleFixedInterval does.  On [which of the properties](https://openstudio-sdk-documentation.s3.amazonaws.com/cpp/OpenStudio-2.8.0-doc/model/html/classopenstudio_1_1model_1_1_schedule_fixed_interval.html) do you specify the list of 8760 values?

macumber replied:
Its in the base class `ScheduleInterval`, the methods is `setTimeSeries` https://github.com/NREL/OpenStudio/blob/develop/openstudiocore/src/model/ScheduleInterval.hpp#L72

macumber replied:
TimeSeries class can go back and forth with an OpenStudio::Vector which is basically an array of floats https://github.com/NREL/OpenStudio/blob/develop/openstudiocore/src/utilities/data/TimeSeries.hpp

chriswmackey replied:
Thanks for all of the info @macumber . ScheduleInterval it is!

macumber commented on app/models/energy/ScheduleRuleset.py:
```diff
@@ -0,0 +1,147 @@
+    unit_type: UnitType
+
+
+class YesOrNo (str, Enum):
```
Seems like this would be more generic than schedules

macumber commented on app/models/energy/ScheduleRuleset.py:
```diff
@@ -0,0 +1,147 @@
+
+
+class ScheduleDiscrete(BaseModel):
+    """This Numeric Type allows all only integers within the range to be valid."""
```
Typo 'This Numeric Type allows *all* only integers within the range to be valid`

macumber commented on app/models/energy/ScheduleBase.py:
```diff
@@ -0,0 +1,60 @@
+class DateTime(BaseModel):
+    """DateTime."""
+
+    month: int = Schema(
```
No year?

tanushree04 replied:
I was using the DateTime class in the pollination schema as reference https://github.com/pollination/simulation-service/blob/master/app/models/common/datetime.py. Should I add the year as well?

macumber replied:
I suggest moving this definition to the `app/models/common/datetime.py` file.  I would suggest giving it a year, but you do want to stay consistent with the rest of the schema.  Maybe needs `is_leap_year` too

chriswmackey replied:
@mostaphaRoudsari and I recently had a conversation and realized that we will need different types of schedules for daylight and energy (largely because of E+ limitations with schedules). While I know I originally said we wanted a common datetime, I think we want to keep date and time as separate objects implemented only on energy as @tanushree04 has it implemented now.  Let's just make sure that the type of input is coordinated with how we have other datetimes in ladybug/honeybee, meaning that time starts counting from hour 0:00 to 23:59 (not 1 to 24) or (1 to 60).

chriswmackey replied:
Update: Minute should still not be able to go past 59 but it seems we will have to support hours going to 24 unless we plan to completely break with E+ convention.

mostaphaRoudsari replied:
This implementation is missing `is_leap_year`. We need to do the validation for leap year in date level otherwise you have to do it in several places.

macumber commented on app/models/energy/ScheduleRuleset.py:
```diff
@@ -0,0 +1,147 @@
+
+    interpolate_to_timestep: YesOrNo = YesOrNo.no
+
+    hour: int = Schema(
```
If you had a separate Time type, you could combine hour and minute into a single time field

macumber commented on app/models/energy/ScheduleRuleset.py:
```diff
@@ -0,0 +1,147 @@
+        le=59
+    )
+
+    value_until_time: float = Schema(
```
Is this an array of values or just one?

macumber commented on app/models/energy/ScheduleRuleset.py:
```diff
@@ -0,0 +1,147 @@
+        regex=r'^[\s.A-Za-z0-9_-]*$'
+    )
+
+    schedule_day: List[ScheduleDay]
```
I think ScheduleDay should have a list of values (covering the whole day) and this should just be one ScheduleDay right?

chriswmackey replied:
I think we will want to support sub-hourly schedules and we can handle the conversion from a list of values to this format on the Python end.

macumber commented on app/models/energy/ScheduleRuleset.py:
```diff
@@ -0,0 +1,147 @@
+
+    winter_designday_schedule: List[ScheduleDay]
+
+    schedule_rule: List[ScheduleRule]
```
Should be schedule_rules, since we don't have a compiler we have to be careful about conventions like indicating which fields are plural or singular.

macumber commented on app/models/energy/ScheduleRuleset.py:
```diff
@@ -0,0 +1,147 @@
+
+    schedule_type_limits: ScheduleTypeLimits
+
+    default_day_schedule: List[ScheduleDay]
```
All of these except schedule_rules should be a single ScheduleDay.

macumber commented on app/models/samples/schedule_ruleset.py:
```diff
@@ -0,0 +1,290 @@
+                    'type': 'ScheduleDay',
+                    'name': 'Default Day 1',
+                    'interpolate_to_timestep': 'No',
+                    'hour': 23,
```
In OpenStudio and EnergyPlus the last value is until 24:00

https://github.com/NREL/OpenStudio/blob/develop/openstudiocore/src/openstudio_app/Resources/default/office_default.osm#L171

tanushree04 committed:
Pydantic error.

mostaphaRoudsari commented on app/models/schemas.py:
```diff
@@ -4,7 +4,11 @@
 from uuid import UUID, uuid4
 from datetime import datetime
-from energy.construction import EnergyConstructionOpaque, EnergyConstructionTransparent
+from .app.models.energy.construction import EnergyConstructionOpaque, EnergyConstructionTransparent
```
```suggestion
from app.models.energy.construction import EnergyConstructionOpaque, EnergyConstructionTransparent
```

This change will fix the importing issue.

@tanushree04 `.` means the current directory. When you add it before app it is telling the module to look for an `app` module from where the file is and since it cannot find one it fails. It makes sense to add it before `energy` because the energy module lives in the same folder.

mostaphaRoudsari commented on app/out.txt:
```diff
@@ -0,0 +1,7 @@
+============================= test session starts =============================
```
we should remove this file from the PR and add it to .gitignore

tanushree04 committed:
try

tanushree04 committed:
trY

tanushree04 committed:
try2

tanushree04 committed:
try2

tanushree04 committed:
Schedule schema and tests

tanushree04 committed:
Schedule schema and tests.

chriswmackey commented on app/models/energy/ScheduleRuleset.py:
```diff
@@ -0,0 +1,113 @@
+        regex=r'^[\s.A-Za-z0-9_-]*$',
+    )
+
+    interpolate_to_timestep: bool
```
I feel that there should be a default value here.  It's usually False in honeybee legacy so maybe we should just do it that way.

chriswmackey commented on app/models/energy/ScheduleRuleset.py:
```diff
@@ -0,0 +1,113 @@
+
+    schedule_day: ScheduleDay
+
+    apply_sunday: bool
```
I am a little confused about what happens if I apply two schedule rules for sunday (or any other day of the week). Do we need a validator to make sure that we don't have two rules occupying the same day?

tanushree04 replied:
@chriswmackey In that case the schedule rule earlier in the list of schedule rules would be given priority.

macumber replied:
That is the OpenStudio rule priority logic.  It can be a bit confusing, we have our own schedule GUI to show which rule is active on each day.  Not sure if you plan on implementing any type of calendar viewer in your UI?

chriswmackey replied:
Good to know, Dan.  If we make any calendar GUI, it is going to be developed pretty far down the line.

chriswmackey commented on app/models/energy/ScheduleRuleset.py:
```diff
@@ -0,0 +1,113 @@
+    unit_type: ScheduleUnitType
+
+
+class DayValue(BaseModel):
```
@macumber , I see that the standards gem just stores the day values as a list:
![image](https://user-images.githubusercontent.com/5567574/59862659-2ee67500-9351-11e9-9b07-e27f8df692f9.png)
Do you usually do the conversion of this list into a set of "values until time" under the hood of OpenStudio?  If so, would you also recommend that we do this same conversion on the Honeybee end?

macumber replied:
If you will always use hourly values then sure, its probably easiest to just store a list.

chriswmackey replied:
There will likely be cases where we want sub-hourly values so let's forego this for now and I will write translators from the standards gem values to the schema values.

chriswmackey reviewed (commented):
Its generally looking good @tanushree04 .  I will probably have some more comments as I star writing the Python code.

macumber commented on app/models/energy/ScheduleBase.py:
```diff
@@ -0,0 +1,100 @@
+
+
+class ScheduleDiscrete(BaseModel):
+    """This Numeric Type allows all only integers within the range to be valid."""
```
typo

macumber commented on app/models/energy/ScheduleBase.py:
```diff
@@ -0,0 +1,100 @@
+    )
+
+
+class Time(BaseModel):
```
I suggest moving this definition to the `app/models/common/datetime.py` file

chriswmackey replied:
See my comment above.  It was aspirational to have a common datetime object but, after thinking about it more, we are really going to need separate for daylight and energy.

mostaphaRoudsari replied:
I agree with @macumber, datetime object should be in a separate file and not in the Schedule module. This is different from the discussion that we had about using them between daylight and energy model schedules.

macumber commented on app/models/energy/ScheduleBase.py:
```diff
@@ -0,0 +1,100 @@
+    )
+
+
+class DateTime(BaseModel):
```
Why not just use the definition in `app/models/common/datetime.py`?

tanushree04 replied:
I was using the `time` class by itself in https://github.com/ladybug-tools-in2/energy-model-schema/blob/fe50743c68915d1fe78579c2437f95e1c4626241/app/models/energy/ScheduleRuleset.py#L38

chriswmackey replied:
I support how @tanushree04  is using it now but I think we can get rid of the DateTime object and have two separate objects of Dates and Times in the ScheduleRuleset.  In other words, Schedule Rules should only be able to accept start and end Dates and not start and end DateTimes.  As I said above, having a common DateTime object would be nice but these schedules do not seem to be the right place for it.

chriswmackey replied:
Update: I realized that, even if we have the date and time objects separate, we can still benefit from using Python's builtin `datetime` module to check the inputs here.  This is because there are separate objects for `date` and `time` within the within the `datetime` module.  So, all that you should have to do, @tanushree04 , to make sure that the user has input a valid date is this:
```
datetime.date(year=2016, day=day, month=month)
```
and this to test if the time is valid:
```
datetime.time(hour=hour, minute=minute)
```
The one exception that we will have to catch is when someone puts in 24:00 for time since it seems that we need to support this in order to deal with EnergyPlus's quirks (see the comments below).

chriswmackey commented on app/models/energy/ScheduleBase.py:
```diff
@@ -0,0 +1,100 @@
+        0,
+        description='A value for hour between `0`-`24`. Default is `0`.',
+        ge=0,
+        le=24,
```
We do not follow EnergyPlus's convention where time starts counting from 1 and goes to 24.  We did this in legacy and realized that it is a huge mistake.  Now we start counting from 0:00 and we go to 23:59.  So this should be from 0 to 23.

macumber replied:
@chriswmackey I agree that time starts at 0:00 each day (it is that way in E+ and OpenStudio too).  However, I do think that the way schedules are specified by the end time (value is until time) suggest that you would want to report the value goes until 24:00 (which is really 0:00 of the next day).  For example:

`value: 1 until: 24:00` would apply the value 1 all times of the day

`value: 1 until: 23:59` would apply the value 1 until 11:59 PM and then have an undefined value for the last minute of the day

it really is an artifact of the way that the values are over an interval, and the interval is specified by the end time of the interval.

chriswmackey replied:
Thanks for clarifying, @macumber. So, am I correct in understanding that E+ will throw and error if someone inputs `value until 24:01` and, if we wanted to support formatting similar to E+, we need to have an extra check to make sure that the user does not do this?

chriswmackey replied:
Does this also mean that E+ expects the end time of a previous interval and the start time of the next interval to be the same? If so, what value does E+ actually use for this timestep that is shared by both intervals (the previous one or the next one)? This lack of clarity makes me think that we might not want to follow this convention just because it is implicit rather than explicit and it is not really in line with the convention that we are following across the rest of Ladybug + Honeybee. Maybe we should just handle the conversion from real time to this E+ overalpped time in the measure. Or maybe we allow it but we need to have methods that make it very clear to the user what is goig on. @mostapha, what do you think here?

macumber replied:
Correct, value until 24:01 is an error, custom validation can say if hr = 24 then min must = 0

The E+ schedule convention is not that bad, E+ is simulating energy over discrete time intervals so it just needs to know when the interval begins and ends

![image](https://user-images.githubusercontent.com/1327850/60359297-eb65b980-9995-11e9-831d-0092ccf77207.png)

chriswmackey replied:
Ok, I guess that we will have to support 24:00 as a valid time in order to deal with this quirk of EnergyPlus.  This also means that we will have to add a validator in the Time object of the schema that permits 24:00 but not anything after (see my comment above about using Python's builtin `datetime.time` class for this).

I just feel the need to state that the way EnergyPlus schedules work is not intuitive and is not the way that we handle timeseries in the rest of Ladybug.  Most users think that, when a schedule says that people occupy an office from 9:00 unit 17:00, they expect that people will be in the office at 9:00 but that they might not be there precisely at 17:00 because they are only supposed to be there until 17:00 (right at 17:00, they should leave).  This is completely against how EnergyPlus schedules and climate data work and I realize that much of this goes back to poor design decisions made about the EPW format over 20 years ago.  But, while non-intuitive design is bad, it is a greater sin to be unclear with the user about what is actually happening in the simulation.  So, for the schema, we will forego the idea of having a `time until` that is a minute less than the next timestep.  We will set the `time until` for the previous interval to be the same as `time from` in the next interval.  And we will work to be as clear with the user as we can that their schedule inputs are likely not what they expect.

This includes several docstrings noting that that the `time from` part of the interval is never actually considered a part of the interval by EnergyPlus (only the timestep after the `time from`). Likewise, `time until` docstrings should be clear that the interval is not actually until that timesetp but rather it is until the timestep following the `time until`.
On the honeybee_energy Python end, we will have to make methods that show exactly what the schedule means timestep-by-timestep so that  users can be made aware of the exact value used at each datetime.  Lastly, on the Python end, we will have to continue the practice that we currently use in Ladybug for timeseries EPW data in that we take the last timestep of the data at the fictional time of 24:00 and move it to be 0:00 for the first timestep of the data.  We will have to do this for all hourly and sub-hourly simulation results of EnergyPlus if we have any hope of correctly aligning it with other timeseries information like solar positions, the output of daylight simulations, or just our own clocks.

chriswmackey commented on app/models/energy/ScheduleBase.py:
```diff
@@ -0,0 +1,100 @@
+        0,
+        description='A value for month between `0`-`60`. Default is `0`.',
+        ge=0,
+        le=60,
```
People should not be able to put in times like 1:60.  A time like that should be 2:00. So this should be from 0 to 59.

chriswmackey commented on app/models/energy/ScheduleFile.py:
```diff
@@ -0,0 +1,59 @@
+from app.models.energy.ScheduleBase import ScheduleContinuous, ScheduleDiscrete, ScheduleNumericType, ScheduleUnitType
+
+
+class ScheduleFile(BaseModel):
```
I think we are also going to need an input for `is_leap_year` here in order to translate it correctly to E+. We should also add a check to ensure that someone specifies the correct number of values depending on whether it is a leap year or not.

tanushree04 commented on app/models/energy/ScheduleFile.py:
```diff
@@ -0,0 +1,59 @@
+        regex=r'^[\s.A-Za-z0-9_-]*$',
+    )
+
+    start_day_of_year: int = Schema(
```
@chriswmackey should I add the start_day_of_year as start_month and start_day the way it is described in OpenStudio?

macumber replied:
We usually think of start_day_of_year as the number of days from 1/1 (starting at one), so July 4th
could be day of year 185

https://www.esrl.noaa.gov/gmd/grad/neubrew/Calendar.jsp

chriswmackey replied:
If the ScheduleFixedInterval supports it, this should be the `Date` object noted above rather than a `start_day_of_year`.  However, I realize that just the Date alone is not enough to know whether the values are for a leap year or not. So @macumber , does the `ScheduleInterval` have a property to denote whether it is for a leap year in the same way that `ScheduleFile` knows that it is a leap year because there are 8784 values in the fie?  Along these lines, it would be really helpful to see some sample code for the `ScheduleInterval`.  I have not yet been able to successfully implement one myself and I still don't quite understand how they are ultimately formatted into the IDF.

macumber replied:
It is a bit weird to work with but I can help with the OpenStudio pieces.  

They are currently translated to `Schedule:Compact`, https://github.com/NREL/OpenStudio/blob/develop/openstudiocore/src/energyplus/ForwardTranslator/ForwardTranslateScheduleFixedInterval.cpp#L75.  However, this might change, we have thoughts of translating these to `Schedule:File` instead, https://github.com/NREL/OpenStudio/issues/50

Here is some example code using this schedule type from the radiance measure

```
       originalLightsValues = lightsTimeseries.values
        lightsValues = OpenStudio::Vector.new(averageIlluminances.size)
        averageIlluminances.each_index do |i|
          dimmingResponse = [(daylightSetpoint - averageIlluminances[i]) / daylightSetpoint, 0].max
          lightsValues[i] = dimmingResponse * originalLightsValues[i]
        end

        # get max lighting power
        lightingLevel = OpenStudio.maximum(lightsValues)

        if lightingLevel <= 0.0
          print_statement("Thermal Zone '#{thermalZone.name}' lighting power is less than or equal to 0, skipping", runner)
          next
        end

        print_statement("Thermal Zone '#{thermalZone.name}' lighting power is: #{lightingLevel.round(0)} W", runner)

        # normalize lights values
        averageIlluminances.each_index do |i|
          lightsValues[i] = lightsValues[i] / lightingLevel
        end

        startDate = OpenStudio::Date.new(OpenStudio::MonthOfYear.new(1), 1)
        interval = OpenStudio::Time.new(0, 1, 0)
        timeseries = OpenStudio::TimeSeries.new(startDate, interval, lightsValues, 'W')
```

macumber replied:
Here is a script you can play with to look at the ScheduleFixedInterval class and translation to IDF.  Just save it run it with the CLI `openstudio test_schedule_file.rb`:

```
is_leap_year = false

if is_leap_year
  num_samples = 8784
else
  num_samples = 8760
end

values = OpenStudio::Vector.new(num_samples)
(0...num_samples).each do |i|
  values[i] = i % 24
end

model = OpenStudio::Model::Model.new
yd = model.getYearDescription
yd.setIsLeapYear(is_leap_year)

startDate = OpenStudio::Date.new(OpenStudio::MonthOfYear.new(1), 1, yd.assumedYear)
interval = OpenStudio::Time.new(0, 1, 0)
timeseries = OpenStudio::TimeSeries.new(startDate, interval, values, 'W')


schedule = OpenStudio::Model::ScheduleInterval::fromTimeSeries(timeseries, model)

puts model

ft = OpenStudio::EnergyPlus::ForwardTranslator.new
idf = ft.translateModel(model)

puts idf
```

chriswmackey replied:
Thanks for posting the sample code, @macumber .  I got the chance to run it and it has definitely given me some perspective.  I was initially thinking that the `FixedInterval` schedule was better suited to our needs but now I see it has many more issues than I realized.  The bug in the github issue is a major one that would definitely the deal-breaker if it still exists but, on top of that I see that just a single year-long compact schedule is so big that it can't be opened in IDFEditor.  Also, all of the extra comments and E+ tags means that the schedule file size is almost 100 times that of its CSV counterpart.  So, altogether, I am positive now that we want to be using ScheduleFile in the measure.  This brings up the question of whether we should stick with the naming convention of ScheduleFixedInterval in the schema and, after thinking about it, I feel that this is still appropriate since we ask the user for a list of values rather than a file.  So we can keep the name but, knowing that there will be using ScheduleFile on the measure end, I am ok if we wanted to stick with the convention of using the start_day_of_year instead of inputting a whole date object.  If there's a datetime module in Ruby, though, it would be nice if we can accept the date object and then convert it to a start_day_of_year in the measure.

On a side note, I had no idea that the year description had all of these capabilities.  The year description object isn't accessible from C# bindings (I have spent quite a few hours in the past trying to access it without any success).  For the sake of the legacy plugins, do you think that the ability to set the leap year will ever be exposed on the C# bindings?

macumber replied:
We have on our list that we should translate schedule interval to schedule file in e+, that is a bug/improvement we could make on the os side.  I think the year description should be available as a method in the Model class, I can try it out sometime when I’m at my computer

chriswmackey replied:
Thanks for the quick response @macumber .  If the change over to ScheduleFile is something that you would be able to make within the IN2 timeline, then I agree that sounds like the best solution.  I know that we budgeted some hours for you to make OpenStudio edits in the IN2 budget and this purpose definitely seems worthy of using those hours.
Investigating the YearDescription in C# bindings is a lower priority since we have workarounds for it in the legacy plugins.  But it still would be really nice to get to at some point.

macumber replied:
I think I can do it for IN2, its good for everyone

tanushree04 commented on app/models/energy/ScheduleFile.py:
```diff
@@ -0,0 +1,59 @@
+        description='Used to define the start day for the analysis period.'
+    )
+
+    values: List[int] = Schema(
```
@chriswmackey  Should I add the `Interval Length`  and `value` instead of a list of values for `ScheduleFixedInterval`?

tanushree04 committed:
-Validation for date and time.

tanushree04 committed:
-ScheduleFixedInterval, validation for Date, Time.

tanushree04 committed:
- Validation for time after 24:00.

mostaphaRoudsari commented on app/models/samples/schedule_ruleset.py:
```diff
@@ -697,3 +697,5 @@
     ]
 }
+
+print(schedule_ruleset_1['schedule_rules'][0])
```
We should not have any print statements like this in a module.

tanushree04 committed:
- Validation for time after 24:00.

mostaphaRoudsari commented on app/models/energy/ScheduleBase.py:
```diff
@@ -0,0 +1,101 @@
+        try: 
+            datetime.time(v, values['hour'], 0)
+        except ValidationError:
+            raise ValidationError('{}/{} is not a valid time.'.format(values['hour'], v))
```
you probably want `{}:{}` not `{}\{}`.

mostaphaRoudsari commented on app/models/energy/ScheduleBase.py:
```diff
@@ -0,0 +1,101 @@
+            return v
+
+    @validator('minute')
+    def check_time_after24(cls, v, values):
```
This is not necessary. The check_time validator already catches this issue.

mostaphaRoudsari commented on app/models/energy/ScheduleBase.py:
```diff
@@ -0,0 +1,101 @@
+        return v     
+
+class ScheduleContinuous(BaseModel):
+    """This Numeric Type allows all numbers, including fractional amounts, within the range to
```
If you have a long docstring like this you can break it down. Also this docstring doesn't say at all that this is a schedule.

See here for how to write a PEP8 compatible docstring.

https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

mostaphaRoudsari commented on app/models/energy/ScheduleBase.py:
```diff
@@ -0,0 +1,101 @@
+"""Schedule Base Schema"""
```
module names most be all lowercase. This filename should be `schedulebase.py`.

mostaphaRoudsari commented on app/models/energy/ScheduleFixedInterval.py:
```diff
@@ -0,0 +1,63 @@
+"""Schedule Fixed Interval Schema"""
```
module names most be all lowercase. This filename should be `schedulefixedinterval.py`. 
if you want to avoid long module names you can create a folder for schedule. copy all the schedule files inside the folder and then name them as `fixedinterval.py` and so on.

mostaphaRoudsari commented on app/models/energy/ScheduleRuleset.py:
```diff
@@ -0,0 +1,132 @@
+"""Schedule Ruleset Schema"""
```
Same comment. module name should be all lowercase.

mostaphaRoudsari commented on app/models/energy/ScheduleBase.py:
```diff
@@ -0,0 +1,101 @@
+
+    schedule_discrete: int = None
+
+class ScheduleNumericType (BaseModel):
```
PEP8 👮 🚨  > Two empty lines before class! Use `autopep8` package in VSCode to fix PEP8 issues like this.

mostaphaRoudsari commented on app/models/energy/ScheduleFixedInterval.py:
```diff
@@ -0,0 +1,63 @@
+from enum import Enum
+from uuid import UUID, uuid4
+import datetime
+from app.models.energy.ScheduleBase import ScheduleContinuous, ScheduleDiscrete, ScheduleNumericType, ScheduleUnitType, Date, Time
```
It looks like you are not using most of these objects. You should only import what is going to be used.

mostaphaRoudsari commented on app/models/energy/ScheduleFixedInterval.py:
```diff
@@ -0,0 +1,63 @@
+
+    start_date: Date
+
+    is_leap_year: bool = Schema(
```
This should also go to date otherwise you will have issues in supporting leap_year. Right now the Date is set to 2016 which I assume is not a leap year.

mostaphaRoudsari commented on app/models/energy/ScheduleFixedInterval.py:
```diff
@@ -0,0 +1,63 @@
+        description='A list of hourly values for the simulation.'
+    )
+
+    @validator('values', whole=True)
```
This validation and the next one should be compiled into one. You can check if items are in range and that will be it.

mostaphaRoudsari commented on app/models/energy/ScheduleFixedInterval.py:
```diff
@@ -0,0 +1,63 @@
+
+    @validator('values', whole=True)
+    def check_max_items(cls, v, values):
+        if 'is_leap_year' == 'True' and len(v) > 8784:
```
This has a number of issues. You are checking two strings against each other which will never be True.

```suggestion
        if values['is_leap_year'] == True and len(v) > 8784:
```

mostaphaRoudsari commented on app/models/energy/ScheduleFixedInterval.py:
```diff
@@ -0,0 +1,63 @@
+        if 'is_leap_year' == 'True' and len(v) > 8784:
+            raise ValueError(
+                'Number of values for leap year can not be more than `8784`.')
+        elif 'is_leap_year' == 'False' and len(v) > 8760:
```
Same here!

```suggestion
        elif values['is_leap_year'] == False and len(v) > 8760:
```

mostaphaRoudsari commented on app/models/energy/ScheduleFixedInterval.py:
```diff
@@ -0,0 +1,63 @@
+        "Ensure the number of values are not less than 24."
+        if len(v) < 24:
+            raise ValueError(
+                'Number of values must be atleast 24.'
```
```suggestion
                'Number of values must be at least 24.'
```

mostaphaRoudsari commented on app/models/energy/ScheduleFixedInterval.py:
```diff
@@ -0,0 +1,63 @@
+        return v
+
+    @validator('is_leap_year')
+    def check_leap_year(cls, v, values):
```
This is the wrong place to do this check. As I said this should go to date itself and you won't need the extra check here.

mostaphaRoudsari commented on app/models/energy/ScheduleRuleset.py:
```diff
@@ -0,0 +1,132 @@
+from enum import Enum
+from uuid import UUID, uuid4
+from datetime import datetime
+from app.models.energy.ScheduleBase import ScheduleContinuous, ScheduleDiscrete, ScheduleNumericType, ScheduleUnitType, Date, Time
```
Again so many modules here that are not used.

mostaphaRoudsari commented on app/models/energy/ScheduleRuleset.py:
```diff
@@ -0,0 +1,132 @@
+    
+
+
+class ScheduleRuleset (BaseModel):
```
```suggestion
class ScheduleRuleset(BaseModel):
```

mostaphaRoudsari commented on app/models/energy/ScheduleRuleset.py:
```diff
@@ -0,0 +1,132 @@
+    start_period: Date
+
+    end_period: Date
+    
```
Too many white lines. PEP8 for the win!

mostaphaRoudsari commented on tests/test_schedule.py:
```diff
@@ -0,0 +1,75 @@
+    ScheduleRuleset.parse_obj(schedule_ruleset_1)
+
+
+def test_schedule_rule_wrong():
```
Each test should ideally check one mistake. This is getting really long. You can break it down into several tests.

mostaphaRoudsari reviewed (commented):
@tanushree04, It looks good overall. My comments fall under two main categories:

1. PEP8: (https://www.python.org/dev/peps/pep-0008/). We follow PEP8 in naming convention, docstrings, white-lines, etc. You can use autopep8 package to automate checking and fixing your files.

2. leap year handling: I think you are better off do the checks for leap_year when you create the date otherwise you have to do these check in several places all over the code.

tanushree04 committed:
Updated validation tests.

tanushree04 committed:
Update schemas.py

tanushree04 committed:
Renamed files.

tanushree04 committed:
To do: Add validation for 24:00.

tanushree04 commented on app/models/energy/schedulefixedinterval.py:
```diff
@@ -0,0 +1,44 @@
+        regex=r'^[\s.A-Za-z0-9_-]*$',
+    )
+
+    start_date: Date
```
@chriswmackey I believe Ruby has a datetime module. Should I keep leave the start_date the way it is? 
The `@validator` in `datetime.py` ensures the start date is not 02/29 if `is_leap_year == False`.

macumber replied:
This allows each schedule in the model to have a different start date.  Difficult situations could arise like:

- Schedule A starts on 1/1 and Schedule B starts on 6/1
- Schedule A starts on 1/1 of a leap year and Schedule B starts on 1/1 of a non leap year
- Schedule A starts at 6/1 but simulation starts on 1/1

Maybe there could be model level validation that all schedules have the same start date?  Or some way to adjust all schedules to the simulation period on translation (filling with 0's as needed)?

macumber replied:
Should this object have an interpolate_to_timestep field?

tanushree04 replied:
I'm not sure about adding the `interpolate_to_timestep`  field because all values entered are for a 24 hour period.

macumber replied:
Interpolate to timestep is really about if values are interpolated between the until times or not.  Here is the relevant field in E+ https://bigladdersoftware.com/epx/docs/8-2/input-output-reference/group-schedules.html#field-interpolate-to-timestep

chriswmackey replied:
@macumber ,
We will have some validation methods for the cases that you mention on the Python end so I would not get too concerned about it here in the schema.  However, this does affect how you will implement  `ScheduleFixedInterival` to use CSV in OpenStudio SDK since (as I recall) the E+ `ScheduleFile` only accepts a full 8760 values in the CSV.  So you will probably have to fill some spaces in the CSV with 0's to get it to work.

tanushree04 committed:
Added validation for 24:00.

mostaphaRoudsari commented on app/models/common/datetime.py:
```diff
@@ -31,15 +31,17 @@ class Date(BaseModel):
 
     @validator('day')
     def check_date(cls, v, values):
+        if not 'month' in values:
```
I'm not sure if this will work.

You are checking the date against 2016 and 2017 without considering the leap year value at all. Did you test this for leap year versus a normal year?

chriswmackey replied:
Yes, you can see that she tests for it in the code below this.

tanushree04 committed:
Validation for date.

mostaphaRoudsari commented on app/models/common/datetime.py:
```diff
@@ -23,26 +29,23 @@ class Date(BaseModel):
-
     @validator('day')
     def check_date(cls, v, values):
+        print(values)
```
?

mostaphaRoudsari commented on app/models/samples/schedule_fixed_interval.py:
```diff
@@ -15,9 +15,9 @@
         'month': 1,
-        'day': 1
+        'day': 1,
+        'is_leap_year': 'False',    
```
This should be False not 'False'

tanushree04 committed:
Updated files.

macumber commented on app/models/energy/scheduleruleset.py:
```diff
@@ -0,0 +1,132 @@
+    value_until_time: float
+
+
+class ScheduleDay(BaseModel):
```
This could be a ScheduleFixed interval with 24 values but I think this way is better (it is more compact and restricts to 24 values only).

macumber commented on app/models/energy/scheduleruleset.py:
```diff
@@ -0,0 +1,132 @@
+    end_period: Date
+
+
+class ScheduleRuleset(BaseModel):
```
@chriswmackey you would probably want some python methods for this class like:

```
schedule_day = schedule_ruleset.getActiveScheduleDay(date)
value = schedule_day.getValue(time)
```

chriswmackey replied:
Agreed. I will definitely implement methods like this on the Python end.

macumber commented on app/models/samples/schedule_fixed_interval.py:
```diff
@@ -0,0 +1,23 @@
+        'day': 1,
+        'is_leap_year': False,    
+    },
+    'values': [20,	20,	20,	20,	20,	20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20
```
Weird formatting here

macumber commented on app/models/schemas.py:
```diff
@@ -24,6 +26,16 @@ class Parent(BaseModel):
         regex=r'^[.A-Za-z0-9_-]*$'
     )
 
+    schedule_ruleset: ScheduleRuleset = Schema(
```
Is `Parent` really `Zone`?  Why does zone have a schedule?

macumber replied:
May want an `AnnualSchedule` type that is the union of `ScheduleRuleset` and `ScheduleFixedInterval` that can be referenced wherever either type can be used?

tanushree04 replied:
I will be clubbing the schedules with the load definitions in the program set. I could add an `AnnualSchedule` there instead of `Union[ScheduleRuleset, ScheduleFixedInterval]`.

macumber commented on test_schedule_file.py:
```diff
@@ -0,0 +1,80 @@
+"""Schedule Fixed Interval Schema"""
```
Is this a test file?  I don't see any assertions, also odd that `Date` and other classes are being redefined here?

tanushree04 committed:
Update schedule_fixed_interval.py

tanushree04 committed:
Updates schedulefixedinterval.py, schema.py

tanushree04 committed:
Changes.
Merge branch 'pr-Schedules' of github.com:ladybug-tools-in2/energy-model-schema into pr-Schedules

chriswmackey reviewed (approved):
Let's just merge this in.  It is good enough right now and, if there is something that we realize that we want to change, we can always do it later.

tanushree04 committed:
Merged changes

tanushree04 committed:
Merge conflicts.

tanushree04 committed:
Merge Conflicts.
